util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,216 @@
+// Normally, the registered ProfileConsumer will receive the cpu profile data per second.
+// If the ProfileConsumer (channel) is full, the latest cpu profile data will not be sent to it.
+// This function is thread-safe.
+func Register(ch ProfileConsumer) {
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs             map[ProfileConsumer]struct{}
+	notifyRegister chan struct{}
+
+	lastDataSize int
+
+	started bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+	wg      sync.WaitGroup
+}
+
+// newParallelCPUProfiler crate a new parallelCPUProfiler.
+func newParallelCPUProfiler() *parallelCPUProfiler {
+	return &parallelCPUProfiler{
+		cs:             make(map[ProfileConsumer]struct{}),
+		notifyRegister: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler is already started")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	p.Lock()
+	if p.started {
+		p.Unlock()
+		return errProfilerAlreadyStarted
+	}
+
+	p.started = true
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+	p.Unlock()
+	p.wg.Add(1)
+	go util.WithRecovery(p.profilingLoop, nil)
+
+	logutil.BgLogger().Info("parallel cpu profiler started")
+	return nil
+}
+
+func (p *parallelCPUProfiler) stop() {
+	p.Lock()
+	if !p.started {
+		p.Unlock()
+		return
+	}
+	p.started = false
+	if p.cancel != nil {
+		p.cancel()
+	}
+	p.Unlock()
+
+	p.wg.Wait()
+	logutil.BgLogger().Info("parallel cpu profiler stopped")
+}
+
+func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	p.cs[ch] = struct{}{}
+	p.Unlock()
+
+	select {
+	case p.notifyRegister <- struct{}{}:
+	default:
+	}
+}
+
+func (p *parallelCPUProfiler) unregister(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	delete(p.cs, ch)
+	p.Unlock()
+}
+
+func (p *parallelCPUProfiler) profilingLoop() {
+	checkTicker := time.NewTicker(defProfileDuration)
+	defer func() {
+		checkTicker.Stop()
+		p.wg.Done()
+	}()
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-p.notifyRegister:
+			p.doProfiling()
+		case <-checkTicker.C:
+			p.doProfiling()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) doProfiling() {
+	if p.consumersCount() == 0 {
+		return
+	}
+
+	data := &ProfileData{Data: bytes.NewBuffer(make([]byte, 0, p.lastDataSize)), Begin: time.Now()}
+	err := pprof.StartCPUProfile(data.Data)
+	if err != nil {
+		data.Error = err
```

Comment:
I'm not sure, but maybe both `data.Begin` and `data.End` doesn't need to exist?

Reply:
great, already removed.